collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java

```diff
@@ -1284,70 +1250,47 @@ private View createViewForFormEnd(FormController formController) {
                     }
                 }
             }
+
             if (saveName == null) {
-                // last resort, default to the form title
-                saveName = formController.getFormTitle();
-            }
-            // present the prompt to allow user to name the form
-            TextView sa = endView.findViewById(R.id.save_form_as);
-            sa.setVisibility(View.VISIBLE);
-            saveAs.setText(saveName);
-            saveAs.setEnabled(true);
-            saveAs.setVisibility(View.VISIBLE);
-            saveAs.addTextChangedListener(new TextWatcher() {
-                @Override
-                public void afterTextChanged(Editable s) {
-                    saveName = String.valueOf(s);
-                }
+                saveName = formSaveViewModel.getFormName();
+            }
+        }
 
-                @Override
-                public void beforeTextChanged(CharSequence s, int start, int count, int after) {
-                }
+        FormEndView endView = new FormEndView(this, formSaveViewModel.getFormName(), saveName, InstancesDaoHelper.isInstanceComplete(true), new FormEndView.Listener() {
+            @Override
+            public void onSaveAsChanged(String saveAs) {
+                saveName = saveAs;
+            }
 
-                @Override
-                public void onTextChanged(CharSequence s, int start, int before, int count) {
+            @Override
+            public void onSaveClicked(boolean markAsFinalized) {
+                if (saveName.length() < 1) {
+                    showShortToast(R.string.save_as_error);
+                } else {
+                    formSaveViewModel.saveForm(getIntent().getData(), markAsFinalized, saveName, true);
                 }
-            });
-        } else {
+            }
+        });
+
+        if (!(boolean) AdminSharedPreferences.getInstance().get(AdminKeys.KEY_MARK_AS_FINALIZED)) {
```

Comment:
I'd like to move more of the setup of this view into the class itself, but I think when we touch it next we can look at adding testing and doing that.